ompi/mca/coll/basic/coll_basic_alltoallw.c

```diff
@@ -31,100 +31,119 @@
 #include "mpi.h"
 #include "ompi/constants.h"
 #include "ompi/datatype/ompi_datatype.h"
+#include "opal/datatype/opal_convertor_internal.h"
 #include "ompi/mca/coll/coll.h"
 #include "ompi/mca/coll/base/coll_tags.h"
 #include "ompi/mca/pml/pml.h"
 
-
+/*
+ * We want to minimize the amount of temporary memory needed while allowing as many ranks
+ * to exchange data simultaneously. We use a variation of the ring algorithm, where in a
+ * single step a process echange the data with both neighbors at distance k (on the left
+ * and the right on a logical ring topology). With this approach we need to pack the data
+ * for a single of the two neighbors, as we can then use the original buffer (and datatype
+ * and count) to send the data to the other.
+ */
 static int
 mca_coll_basic_alltoallw_intra_inplace(const void *rbuf, const int *rcounts, const int *rdisps,
                                        struct ompi_datatype_t * const *rdtypes,
                                        struct ompi_communicator_t *comm,
                                        mca_coll_base_module_t *module)
 {
-    int i, j, size, rank, err = MPI_SUCCESS, max_size;
-    ompi_request_t *req;
-    char *tmp_buffer, *save_buffer = NULL;
-    ptrdiff_t ext, gap = 0;
-
-    /* Initialize. */
+    int i, size, rank, left, right, err = MPI_SUCCESS;
+    ompi_request_t *req = MPI_REQUEST_NULL;
+    char *tmp_buffer = NULL;
+    size_t max_size = 0, packed_size, msg_size_left, msg_size_right;
+    opal_convertor_t convertor;
 
     size = ompi_comm_size(comm);
-    rank = ompi_comm_rank(comm);
-
-    /* If only one process, we're done. */
-    if (1 == size) {
+    if (1 == size) {  /* If only one process, we're done. */
         return MPI_SUCCESS;
     }
+    rank = ompi_comm_rank(comm);
 
-    /* Find the largest receive amount */
-    for (i = 0, max_size = 0 ; i < size ; ++i) {
-        ext = opal_datatype_span(&rdtypes[i]->super, rcounts[i], &gap);
-
-        max_size = ext > max_size ? ext : max_size;
+    /* Find the largest amount of packed send/recv data amoing all peers where
+     * we need to pack before the send.
+     */
+    for (i = 1 ; i <= (size >> 1) ; ++i) {
+        right = (rank + i) % size;
+        ompi_proc_t *ompi_proc = ompi_comm_peer_lookup(comm, right);
+
+        if( opal_local_arch == ompi_proc->super.proc_convertor->master->remote_arch) {
+            opal_datatype_type_size(&rdtypes[right]->super, &packed_size);
+        } else {
```

Comment:
minor optimization: we only need the if/else clause if heterogeneous support is enabled, right?

Reply:
Without heterogeneous support we only need to call the opal_datatype_type_size. But the code gets ugly, and the it only saves us an if branch (one that will be mostly taken). I could have marked it as `likely`.

Reply:
strictly speaking, that's one branch plus one lookup (e.g. no need for `ompi_proc` unless heterogeneous support).

Otherwise, that looks fair to me.

Reply:
You're right, let's save the lookup. I have now pushed the INPLACE versions for MPI_Alltoall, MPI_Alltoallv and MPI_Alltoallw. Please review.